### PR TITLE
[metrics 4/x] Metrics exporter rules

### PR DESCRIPTION
**Pull request by zeeke**
PrometheusRules allow recording pre-defined queries. Use
`sriov_kubepoddevice` metric to add `pod|namespace` pair
to the sriov metrics.

Here is an example of the raw exported metrics:
```
sriov_kubepoddevice{container="testpmd",dev_type="openshift.io/inteldpdk",namespace="cnf-4916",pciAddr="0000:17:01.4",pod="dpdk-intel-client"} 1

sriov_vf_rx_broadcast{numa_node="0",pciAddr="0000:17:01.4",pf="ens2f0",vf="4"} 1.0926018e+07
sriov_vf_rx_bytes{numa_node="0",pciAddr="0000:17:01.4",pf="ens2f0",vf="4"} 7.83952134e+08
sriov_vf_rx_dropped{numa_node="0",pciAddr="0000:17:01.4",pf="ens2f0",vf="4"} 0
sriov_vf_rx_multicast{numa_node="0",pciAddr="0000:17:01.4",pf="ens2f0",vf="4"} 0
sriov_vf_rx_packets{numa_node="0",pciAddr="0000:17:01.4",pf="ens2f0",vf="4"} 1.0926018e+07
sriov_vf_tx_bytes{numa_node="0",pciAddr="0000:17:01.4",pf="ens2f0",vf="4"} 0
sriov_vf_tx_dropped{numa_node="0",pciAddr="0000:17:01.4",pf="ens2f0",vf="4"} 1.0926018e+07
sriov_vf_tx_packets{numa_node="0",pciAddr="0000:17:01.4",pf="ens2f0",vf="4"} 0
```

Proposed prometheus rules allow to query the following new metrics:
- `network:sriov_vf_tx_packets`
- `network:sriov_vf_rx_packets`
- `network:sriov_vf_tx_bytes`
- `network:sriov_vf_rx_bytes`
- `network:sriov_vf_tx_dropped`
- `network:sriov_vf_rx_dropped`
- `network:sriov_vf_rx_broadcast`
- `network:sriov_vf_rx_multicast`

### CODE REVIEW

File: bindata/manifests/metrics-exporter/metrics-prometheus-rule.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
+---
+{{ if .IsPrometheusOperatorInstalled }}
```
Comment (adrianchiris) [inline bindata/manifests/metrics-exporter/metrics-prometheus-rule.yaml:2]: i recall we discussed having a separate env var for deploying rules right ?

from aug 06 2024 meeting notes:
> Adding a new env variable to decide if deploy prometheus rules

Comment (zeeke) [reply bindata/manifests/metrics-exporter/metrics-prometheus-rule.yaml:2]: yes, I'm fixing that

**Review by SchSeba — APPROVED** [review]
LGTM

File: deploy/role.yaml [hunk 2]
```diff
@@ -29,6 +29,7 @@ rules:
   - monitoring.coreos.com
   resources:
   - servicemonitors
+  - prometheusrules
```
Comment (adrianchiris) [inline deploy/role.yaml:32]: Do we support deletion of prometheus objects ?

Comment (adrianchiris) [reply deploy/role.yaml:32]: For when sno is redeployed with prometheus disabled (e.g in helm chart)

Comment (zeeke) [reply deploy/role.yaml:32]: good point. we have to support object deletion as well. I handle it in 
- https://github.com/k8snetworkplumbingwg/sriov-network-operator/pull/769
but it's better to merge that changes into these PR.

**Comment by zeeke** [comment]
@adrianchiris can we move this forward?

**Review by adrianchiris — APPROVED** [review]
LGTM

File: deployment/sriov-network-operator-chart/README.md [hunk 3]
```diff
@@ -89,6 +89,7 @@ We have introduced the following Chart parameters.
 | `operator.metricsExporter.prometheusOperator.enabled` | bool | false | Wheter the operator shoud configure Prometheus resources or not (e.g. `ServiceMonitors`). |
 | `operator.metricsExporter.prometheusOperator.serviceAccount` | string | `prometheus-k8s` | The service account used by the Prometheus Operator. This is used to give Prometheus the permission to list resource in the SR-IOV operator namespace |
 | `operator.metricsExporter.prometheusOperator.namespace` | string | `monitoring` | The namespace where the Prometheus Operator is installed. Setting this variable makes the operator deploy `monitoring.coreos.com` resources. |
+| `operator.metricsExporter.prometheusOperator.deployRules` | boold | false | Whether the operator should deploy `PrometheusRules` to scrape namesapced version of metrics. |
```
[the remark this answers is not in the document]
Comment (adrianchiris) [reply deployment/sriov-network-operator-chart/README.md:92]: @zeeke lets address this and we can merge IMO

**Review by Eoghan1232 — APPROVED** [review]

**Pull request merged**